File: src/main/java/io/rsocket/cli/http2/Http2ClientTransport.java
package io.rsocket.cli.http2;

import io.rsocket.DuplexConnection;
import io.rsocket.cli.HeaderAware;
import io.rsocket.transport.ClientTransport;
import java.net.InetSocketAddress;
import java.net.URI;
import java.util.Collections;
import java.util.Map;
import java.util.logging.Logger;
import org.eclipse.jetty.http2.api.Session;
import org.eclipse.jetty.http2.api.server.ServerSessionListener;
import org.eclipse.jetty.http2.client.HTTP2Client;
import org.eclipse.jetty.util.Promise;
import org.eclipse.jetty.util.log.JavaUtilLog;
import org.eclipse.jetty.util.log.Log;
import org.eclipse.jetty.util.ssl.SslContextFactory;
import org.eclipse.jetty.util.thread.QueuedThreadPool;
import org.eclipse.jetty.util.thread.ScheduledExecutorScheduler;
import reactor.core.publisher.Mono;

public class Http2ClientTransport implements ClientTransport, HeaderAware {
  private static final Logger log = Logger.getLogger(Http2DuplexConnection.class.getName());

  static {
    Log.setLog(new JavaUtilLog());
  }

  private final URI uri;
  // assume simplified header model
  private Map<String, String> headers;

  public Http2ClientTransport(URI uri) {
    this(uri, Collections.emptyMap());
  }

  public Http2ClientTransport(URI uri, Map<String, String> headers) {
    this.uri = uri;
    this.headers = headers;
  }

  @Override
  public void setHeaders(Map<String, String> headers) {
    this.headers = headers;
  }

  @Override
  public Mono<DuplexConnection> connect() {
    return createSession().flatMap(s -> Http2DuplexConnection.create(s, uri, headers));
  }

  private Mono<Session> createSession() {
    return Mono.create(
        c -> {
          HTTP2Client client = new HTTP2Client();
          client.setExecutor(daemonClientExecutor());
          client.setScheduler(daemonClientScheduler());
          SslContextFactory sslContextFactory = null;
          if (uri.getScheme().equals("https")) {
            sslContextFactory = new SslContextFactory();
            client.addBean(sslContextFactory);
          }

          try {
            client.start();

              int port = getPort();
              System.out.println(port);
              client.connect(
                sslContextFactory,
                new InetSocketAddress(uri.getHost(), port),
                new ServerSessionListener.Adapter(),
                new Promise<Session>() {
                  @Override
                  public void succeeded(Session result) {
                    c.success(result);
                  }

                  @Override
                  public void failed(Throwable x) {
                    c.error(x);
                  }
                });
          } catch (Exception e) {
            c.error(e);
          }
        });
  }

  private ScheduledExecutorScheduler daemonClientScheduler() {
    ScheduledExecutorScheduler scheduler = new ScheduledExecutorScheduler("jetty-scheduler", true);
    return scheduler;
  }

  private QueuedThreadPool daemonClientExecutor() {
    QueuedThreadPool executor = new QueuedThreadPool();
    executor.setDaemon(true);
    return executor;
  }

  private int getPort() {
    if (uri.getPort() != -1) {
      return uri.getPort();
    }

    return "https".equals(uri.getScheme()) ? 443 : 80;
  }
}

File: src/main/kotlin/io/rsocket/cli/http2/Http2ClientTransport.kt
package io.rsocket.cli.http2

import io.rsocket.DuplexConnection
import io.rsocket.transport.ClientTransport
import io.rsocket.transport.TransportHeaderAware
import org.eclipse.jetty.http.HttpScheme
import org.eclipse.jetty.http2.api.Session
import org.eclipse.jetty.http2.api.server.ServerSessionListener
import org.eclipse.jetty.http2.client.HTTP2Client
import org.eclipse.jetty.util.Promise
import org.eclipse.jetty.util.log.JavaUtilLog
import org.eclipse.jetty.util.log.Log
import org.eclipse.jetty.util.ssl.SslContextFactory
import org.eclipse.jetty.util.thread.QueuedThreadPool
import org.eclipse.jetty.util.thread.ScheduledExecutorScheduler
import reactor.core.publisher.Mono
import java.net.InetSocketAddress
import java.net.URI
import java.util.function.Supplier

class Http2ClientTransport @JvmOverloads constructor(
        private val uri: URI,
        private var transportHeadersFn: () -> Map<String, String> = { mutableMapOf() }) : ClientTransport, TransportHeaderAware {

    override fun setTransportHeaders(transportHeaders: Supplier<Map<String, String>>?) {
        this.transportHeadersFn = { transportHeaders?.get() ?: mapOf() }
    }

    override fun connect(): Mono<DuplexConnection> =
            createSession().flatMap { s -> Http2DuplexConnection.create(s, uri, transportHeadersFn()) }

    private fun createSession(): Mono<Session> = Mono.create { c ->
        val client = HTTP2Client()
        client.executor = daemonClientExecutor()
        client.scheduler = daemonClientScheduler()
        var sslContextFactory: SslContextFactory? = null
        if (HttpScheme.HTTPS.`is`(uri.scheme)) {
            sslContextFactory = SslContextFactory()
            client.addBean(sslContextFactory)
        }

        try {
            client.start()

            client.connect(
                    sslContextFactory,
                    InetSocketAddress(uri.host, port),
                    ServerSessionListener.Adapter(),
                    object : Promise<Session> {
                        override fun succeeded(result: Session?) {
                            c.success(result)
                        }

                        override fun failed(x: Throwable?) {
                            c.error(x!!)
                        }
                    })
        } catch (e: Exception) {
            c.error(e)
        }
    }

    private fun daemonClientScheduler(): ScheduledExecutorScheduler =
            ScheduledExecutorScheduler("jetty-scheduler", true)

    private fun daemonClientExecutor(): QueuedThreadPool {
        val executor = QueuedThreadPool()
        executor.isDaemon = true
        return executor
    }

    private val port: Int
        get() {
            if (uri.port != -1) {
                return uri.port
            }

            return if (HttpScheme.HTTPS.`is`(uri.scheme)) 443 else 80
        }

    companion object {
        init {
            Log.setLog(JavaUtilLog())
        }
    }
}
